// @loaders.gl/mvt/dist/esm/lib/binary-vector-tile/vector-tile-feature.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { classifyRings, project, readFeature } from '../../helpers/binary-util-functions';
let endPos;
let cmd;
let cmdLen;
let length;
let x;
let y;
let i;
export const TEST_EXPORTS = {
  classifyRings
};
export default class VectorTileFeature {
  static get types() {
    return ['Unknown', 'Point', 'LineString', 'Polygon'];
  }

  constructor(pbf, end, extent, keys, values, firstPassData) {
    _defineProperty(this, "properties", void 0);

    _defineProperty(this, "extent", void 0);

    _defineProperty(this, "type", void 0);

    _defineProperty(this, "id", void 0);

    _defineProperty(this, "_pbf", void 0);

    _defineProperty(this, "_geometry", void 0);

    _defineProperty(this, "_keys", void 0);

    _defineProperty(this, "_values", void 0);

    _defineProperty(this, "_firstPassData", void 0);

    this.properties = {};
    this.extent = extent;
    this.type = 0;
    this.id = null;
    this._pbf = pbf;
    this._geometry = -1;
    this._keys = keys;
    this._values = values;
    this._firstPassData = firstPassData;
    pbf.readFields(readFeature, this, end);
  }

  loadGeometry() {
    const pbf = this._pbf;
    pbf.pos = this._geometry;
    endPos = pbf.readVarint() + pbf.pos;
    cmd = 1;
    length = 0;
    x = 0;
    y = 0;
    i = 0;
    const lines = [];
    const data = [];

    while (pbf.pos < endPos) {
      if (length <= 0) {
        cmdLen = pbf.readVarint();
        cmd = cmdLen & 0x7;
        length = cmdLen >> 3;
      }

      length--;

      if (cmd === 1 || cmd === 2) {
        x += pbf.readSVarint();
        y += pbf.readSVarint();

        if (cmd === 1) {
          lines.push(i);
        }

        data.push(x, y);
        i += 2;
      } else if (cmd === 7) {
        if (i > 0) {
          const start = lines[lines.length - 1];
          data.push(data[start], data[start + 1]);
          i += 2;
        }
      } else {
        throw new Error("unknown command ".concat(cmd));
      }
    }

    return {
      data,
      lines
    };
  }

  _toBinaryCoordinates(transform) {
    let geom = this.loadGeometry();
    transform(geom.data, this);
    const coordLength = 2;

    switch (this.type) {
      case 1:
        this._firstPassData.pointFeaturesCount++;
        this._firstPassData.pointPositionsCount += geom.lines.length;
        break;

      case 2:
        this._firstPassData.lineFeaturesCount++;
        this._firstPassData.linePathsCount += geom.lines.length;
        this._firstPassData.linePositionsCount += geom.data.length / coordLength;
        break;

      case 3:
        const classified = classifyRings(geom);
        this._firstPassData.polygonFeaturesCount++;
        this._firstPassData.polygonObjectsCount += classified.lines.length;

        for (const lines of classified.lines) {
          this._firstPassData.polygonRingsCount += lines.length;
        }

        this._firstPassData.polygonPositionsCount += classified.data.length / coordLength;
        geom = classified;
        break;
    }

    geom.type = VectorTileFeature.types[this.type];

    if (geom.lines.length > 1) {
      geom.type = "Multi".concat(geom.type);
    }

    const result = {
      type: 'Feature',
      geometry: geom,
      properties: this.properties
    };

    if (this.id !== null) {
      result.id = this.id;
    }

    return result;
  }

  toBinaryCoordinates(options) {
    if (typeof options === 'function') {
      return this._toBinaryCoordinates(options);
    }

    return this._toBinaryCoordinates(project);
  }

}
