// @loaders.gl/mvt/dist/es5/lib/binary-vector-tile/vector-tile-feature.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TEST_EXPORTS = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _binaryUtilFunctions = require("../../helpers/binary-util-functions");

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var endPos;
var cmd;
var cmdLen;
var length;
var x;
var y;
var i;
var TEST_EXPORTS = {
  classifyRings: _binaryUtilFunctions.classifyRings
};
exports.TEST_EXPORTS = TEST_EXPORTS;

var VectorTileFeature = function () {
  function VectorTileFeature(pbf, end, extent, keys, values, firstPassData) {
    (0, _classCallCheck2.default)(this, VectorTileFeature);
    (0, _defineProperty2.default)(this, "properties", void 0);
    (0, _defineProperty2.default)(this, "extent", void 0);
    (0, _defineProperty2.default)(this, "type", void 0);
    (0, _defineProperty2.default)(this, "id", void 0);
    (0, _defineProperty2.default)(this, "_pbf", void 0);
    (0, _defineProperty2.default)(this, "_geometry", void 0);
    (0, _defineProperty2.default)(this, "_keys", void 0);
    (0, _defineProperty2.default)(this, "_values", void 0);
    (0, _defineProperty2.default)(this, "_firstPassData", void 0);
    this.properties = {};
    this.extent = extent;
    this.type = 0;
    this.id = null;
    this._pbf = pbf;
    this._geometry = -1;
    this._keys = keys;
    this._values = values;
    this._firstPassData = firstPassData;
    pbf.readFields(_binaryUtilFunctions.readFeature, this, end);
  }

  (0, _createClass2.default)(VectorTileFeature, [{
    key: "loadGeometry",
    value: function loadGeometry() {
      var pbf = this._pbf;
      pbf.pos = this._geometry;
      endPos = pbf.readVarint() + pbf.pos;
      cmd = 1;
      length = 0;
      x = 0;
      y = 0;
      i = 0;
      var lines = [];
      var data = [];

      while (pbf.pos < endPos) {
        if (length <= 0) {
          cmdLen = pbf.readVarint();
          cmd = cmdLen & 0x7;
          length = cmdLen >> 3;
        }

        length--;

        if (cmd === 1 || cmd === 2) {
          x += pbf.readSVarint();
          y += pbf.readSVarint();

          if (cmd === 1) {
            lines.push(i);
          }

          data.push(x, y);
          i += 2;
        } else if (cmd === 7) {
          if (i > 0) {
            var start = lines[lines.length - 1];
            data.push(data[start], data[start + 1]);
            i += 2;
          }
        } else {
          throw new Error("unknown command ".concat(cmd));
        }
      }

      return {
        data: data,
        lines: lines
      };
    }
  }, {
    key: "_toBinaryCoordinates",
    value: function _toBinaryCoordinates(transform) {
      var geom = this.loadGeometry();
      transform(geom.data, this);
      var coordLength = 2;

      switch (this.type) {
        case 1:
          this._firstPassData.pointFeaturesCount++;
          this._firstPassData.pointPositionsCount += geom.lines.length;
          break;

        case 2:
          this._firstPassData.lineFeaturesCount++;
          this._firstPassData.linePathsCount += geom.lines.length;
          this._firstPassData.linePositionsCount += geom.data.length / coordLength;
          break;

        case 3:
          var classified = (0, _binaryUtilFunctions.classifyRings)(geom);
          this._firstPassData.polygonFeaturesCount++;
          this._firstPassData.polygonObjectsCount += classified.lines.length;

          var _iterator = _createForOfIteratorHelper(classified.lines),
              _step;

          try {
            for (_iterator.s(); !(_step = _iterator.n()).done;) {
              var lines = _step.value;
              this._firstPassData.polygonRingsCount += lines.length;
            }
          } catch (err) {
            _iterator.e(err);
          } finally {
            _iterator.f();
          }

          this._firstPassData.polygonPositionsCount += classified.data.length / coordLength;
          geom = classified;
          break;
      }

      geom.type = VectorTileFeature.types[this.type];

      if (geom.lines.length > 1) {
        geom.type = "Multi".concat(geom.type);
      }

      var result = {
        type: 'Feature',
        geometry: geom,
        properties: this.properties
      };

      if (this.id !== null) {
        result.id = this.id;
      }

      return result;
    }
  }, {
    key: "toBinaryCoordinates",
    value: function toBinaryCoordinates(options) {
      if (typeof options === 'function') {
        return this._toBinaryCoordinates(options);
      }

      return this._toBinaryCoordinates(_binaryUtilFunctions.project);
    }
  }], [{
    key: "types",
    get: function get() {
      return ['Unknown', 'Point', 'LineString', 'Polygon'];
    }
  }]);
  return VectorTileFeature;
}();

exports.default = VectorTileFeature;
